docs/source/daml/reference/contract-keys.rst

```diff
@@ -0,0 +1,95 @@
+
+You can use contract keys to stably refer to a contract, even through iterations of instances of it.
+
+Here's an example of setting up a contract key for a bank account, to act as a bank account ID:
+
+.. literalinclude:: ../code-snippets/ContractKeys.daml
+   :language: daml
+   :start-after: -- start contract key setup snippet
+   :end-before: -- end contract key setup snippet
+
+What can be a contract key
+**************************
+
+The key should either be a tuple or a record, and it **must** include every party that you specify as a ``maintainer`` (see `Specifying maintainers`_ below).
+
+For example, with ``maintainer bank``, ``key (bank, number) : (Party, Text)`` is valid; but wouldn't be if you removed ``bank`` from the key.
+
+It's best to use simple types for your keys like ``Text`` or ``Integer``, rather than a list or more complex type.
+
+Specifying maintainers
+**********************
+
+If you specify a contract key for a template, you must also specify a ``maintainer`` or maintainers, in a similar way to specifying signatories or observers. Maintainers are the parties that know about all of the keys that they are party to, and are used by the engine to guarantee uniqueness of contract keys.  The maintainers **must** be signatories or observers of the contract.
+
+Keys are unique only to their maintainers. For example, say you have a key that you're using as the identifer for a ``BankAccount`` contract. You might have ``key (bank, accountId) : (Party, Text)``. When you create a new bank account, the contract key ensures that no-one else can have an account with the same ``accountID`` at that bank. But that doesn't apply to other banks: for a contract with a different bank as maintainer, you could happily re-use that ``accountID``.
+
+When you're writing DAML models, the maintainers only matter since they affect authorization -- much like signatories and observers. You don't need to do anything to "maintain" the keys.
+
+Checking of the keys is done automatically at execution time, by the DAML exeuction engine: if someone tries to create a new contract that duplicates an existing contract key, the execution engine will cause that creation to fail. 
+
+Contract keys functions
+***********************
+
+Contract keys introduce several new functions.
+
+``fetchByKey``
+==============
+
+``fetchedContract <- fetchByKey @ContractType contractKey``
+
+Use ``fetchByKey`` to fetch the ID and data of the contract with the specified key. It is an alternative to the currently-used ``fetch``.
+
+It returns a tuple of the ID and the contract object (containing all its data). 
+
+You need authorization from **at least one** of the maintainers to run ``fetchByKey``. A maintainer can authorize by being a signatory, or by submitting the command/being a controller for the choice.
```

Comment:
I'm not sure we should explain how to authorize something here. We should instead do this in a more general setting because it is relevant for `create` and `exercise` as well. Can we just drop the second sentence for now?

Reply:
I think this should be "as well as" rather than "instead"? I'd like to have the authorization information at the point of need - ie here - but also we could have a more general page explaining how it works overall.

Reply:
The authorization model is not simple enough to repeat it inline wherever it is needed. The explanation we have is not quite accurate. A proper explanation would be too long:
> A maintainer can authorize by being a signatory of the contract containing the choice in whose body the `fetchByKey` is run or by being a controller of this choice.

Reply:
Oh, I see what you mean.